strawberryfields/decompositions.py

```diff
@@ -476,6 +476,190 @@ def rectangular_symmetric(V, tol=1e-11):
+            for j in reversed(range(nsize - 1 - i)):
+                tilist.append(nullMZi(i + j + 1, j, localV))
+                tilist[-1][2] %= (2 * np.pi)
+                tilist[-1][3] %= (2 * np.pi)
+                # repeat modulo operations , otherwise the input unitary
+                # numpy.identity(20) yields an external_phase of exactly 2 * pi
+                tilist[-1][2] %= (2 * np.pi)
+                tilist[-1][3] %= (2 * np.pi)
+                localV = localV @ MZi(*tilist[-1])
+        else:
+            for j in range(nsize - 1 - i):
+                tlist.append(nullMZ(i + j + 1, j, localV))
+                tlist[-1][2] %= (2 * np.pi)
+                tlist[-1][3] %= (2 * np.pi)
+                # repeat modulo operations , otherwise the input unitary
+                # numpy.identity(20) yields an external_phase of exactly 2 * pi
+                tlist[-1][2] %= (2 * np.pi)
+                tlist[-1][3] %= (2 * np.pi)
+                localV = MZ(*tlist[-1]) @ localV
+
+    return tilist, np.diag(localV), tlist
+
+def rectangular_phase_end_MZ(V, tol=1e-11):
+    r"""Rectangular decomposition of a unitary matrix, with all
+    local phase shifts placed after the interferometers.
+
+    See :cite:`clements2016` for more details.
+
+    Final step in the decomposition of a given discrete unitary matrix.
+    The output is of the form given in Eq. 5.
+
+    Args:
+        V (array[complex]): unitary matrix of size n_size
+        tol (float): the tolerance used when checking if the matrix is unitary:
+            :math:`|VV^\dagger-I| \leq` tol
+
+    Returns:
+        tuple[array]: returns a tuple of the form ``(tlist, np.diag(localV), None)``
+            where:
+
+            * ``tlist``: list containing ``[n,m,theta,phi,n_size]`` of the MZ unitaries needed
+            * ``localV``: Diagonal unitary matrix to be applied at the end of circuit
+    """
+    tilist, diags, tlist = rectangular_MZ(V, tol)
+    new_tlist, new_diags = tilist.copy(), diags.copy()
+
+    # Push each beamsplitter through the diagonal unitary
+    for i in reversed(tlist):
+        em, en = int(i[0]), int(i[1])
+        alpha, beta = np.angle(new_diags[em]), np.angle(new_diags[en])
+        phi_i, phi_e = i[2], i[3]
+
+        # The new parameters required for D',MZ' st. MZ^(-1)D = D'MZ'
+
+        new_phi_e = (alpha - beta) % (2 * np.pi)
+        new_alpha = (beta - phi_e - phi_i + np.pi) % (2 * np.pi)
+        new_beta = (beta - phi_i + np.pi) % (2 * np.pi)
+        new_phi_i = phi_i % (2 * np.pi)
+        # repeat modulo operations , otherwise the input unitary
+        # numpy.identity(20) yields an external_phase of exactly 2 * pi
+        new_phi_i %= (2 * np.pi)
+        new_phi_e %= (2 * np.pi)
+        
+        new_i = [i[0], i[1], new_phi_i, new_phi_e, i[4]]
+        new_diags[em], new_diags[en] = np.exp(1j*new_alpha), np.exp(1j*new_beta)
+
+        new_tlist = new_tlist + [new_i]
+
+    return new_tlist, new_diags, None
```

Comment:
Nice! From my understanding, this avoids the added complication from rectangular_symmetric in that it avoids decomposing into SF BSgates, instead decomposing directly into hardware MZgates, and therefore avoiding the process whereby the BSgates and Rgates are modified to transform them back into MZgates?

Reply:
This is spot on except that it avoids decomposing into the `T` matrices from the Clements paper (which are neither SF BS gates nor hardware MZgates).